quic-api.md

```diff
@@ -0,0 +1,179 @@
+- `QuicConnection`: QUIC connection, corresponding to [RFC 9000 -  5. Connections](https://www.rfc-editor.org/rfc/rfc9000.html#name-connections).
+- `QuicStream`: QUIC stream, corresponding to [RFC 9000 -   2. Streams ](https://www.rfc-editor.org/rfc/rfc9000.html#name-streams).
+
+But before any usage of these classes, user code should check whether QUIC is currently supported, as `libmsquic` might be missing, or TLS 1.3 might not be supported. For that, both `QuicListener` and `QuicConnection` expose a static property `IsSupported`:
+```C#
+if (QuicListener.IsSupported)
+{
+    // Use QuicListener
+}
+else
+{
+    // Fallback/Error
+}
+
+if (QuicConnection.IsSupported)
+{
+    // Use QuicConnection
+}
+else
+{
+    // Fallback/Error
+}
+```
+Note that at the moment, both these properties are in sync and will report the same value, but that might change in the future. So we recommend to check [`QuicListener.IsSupported`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener?view=net-7.0) for server-scenarios and [`QuicConnection.IsSupported`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quicconnection.issupported?view=net-7.0) for the client ones.
+
+### QuicListener
+
+[`QuicListener`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener?view=net-7.0) represents a server side class that accepts incoming connections from the clients. The listener is constructed and started with a static method [`QuicListener.ListenAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener.listenasync?view=net-7.0). The method accepts an instance of [`QuicListenerOptions`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclisteneroptions?view=net-7.0) class with all the settings necessary to start the listener and accept incoming connections. After that, listener is ready to hand out connections via [`AcceptConnectionAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener.acceptconnectionasync?view=net-7.0). Connections returned by this method are always fully connected, meaning that the TLS handshake is finished and the connection is ready to be used. Finally to stop listening and release all resources, [`DisposeAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.net.quic.quiclistener.disposeasync?view=net-7.0) must be called.
+
+The sample usage of `QuicListener`:
+```C#
+using System.Net.Quic;
+
+// First, check if QUIC is supported.
+if (!QuicListener.IsSupported)
+{
+    Console.WriteLine("QUIC is not supported, check for presence of libmsquic and support of TLS 1.3.");
+    return;
+}
+
+// We want the same configuration for each incoming connection, so we prepare the connection option upfront and reuse them.
+// This represents the minimal configuration necessary to accept a connection.
+var serverConnectionOptions = new QuicServerConnectionOptions()
+{
+    // Used to abort stream if it's not properly closed by the user.
```

Comment:
Maybe we should add something like "Can be any number as defined by the application level protocol"

Reply:
Instead of that, I'm going with the real H/3 error codes as I'm using "h3" ALPN. I think that concrete example is here better.

Reply:
Used different constant and added "Protocol-dependent error code." comment